File: sdk/typescript/rooch-sdk/src/utils/bitcoin.ts
// Copyright (c) RoochNetwork
// SPDX-License-Identifier: Apache-2.0

import { Bytes } from '../types/index.js'

export function validateWitness(version: number, data: Bytes) {
  if (data.length < 2 || data.length > 40) throw new Error('Witness: invalid length')
  if (version > 16) throw new Error('Witness: invalid version')
  if (version === 0 && !(data.length === 20 || data.length === 32))
    throw new Error('Witness: invalid length for version')
}

File: sdk/golang/rooch-sdk/utils/bitcoin.go
package utils

import "fmt"

// validateWitness checks if the witness program is valid according to BIP-0141 rules
func ValidateWitness(version int, data []byte) error {
    if len(data) < 2 || len(data) > 40 {
        return fmt.Errorf("Witness: invalid length")
    }
    
    if version > 16 {
        return fmt.Errorf("Witness: invalid version")
    }
    
    if version == 0 && !(len(data) == 20 || len(data) == 32) {
        return fmt.Errorf("Witness: invalid length for version")
    }
    
    return nil
} 